zero/src/block_interval.rs

```diff
@@ -1,16 +1,41 @@
 use std::pin::Pin;
 use std::sync::Arc;
+use std::{future::Future, ops::Range};
 
-use alloy::primitives::B256;
-use alloy::rpc::types::eth::BlockId;
-use alloy::{hex, providers::Provider, transports::Transport};
+use alloy::providers::Provider;
+use alloy::rpc::types::BlockTransactionsKind;
+use alloy::rpc::types::{eth::BlockId, Block};
+use alloy::transports::Transport;
 use anyhow::{anyhow, Result};
 use async_stream::try_stream;
 use futures::Stream;
+#[cfg(test)]
+//use mockall::mock;
 use tracing::info;
 
-use crate::parsing;
-use crate::provider::CachedProvider;
+//#[cfg_attr(test, automock)]
+pub trait BlockIntervalProvider<T> {
+    fn get_block_by_id(
+        &self,
+        block_id: BlockId,
+    ) -> impl Future<Output = anyhow::Result<Option<Block>>> + Send;
+
+    fn latest_block_number(&self) -> impl Future<Output = anyhow::Result<u64>> + Send;
+}
+
+impl<T: Transport + Clone, P: Provider<T>> BlockIntervalProvider<T> for P {
```

Comment:
@0xaatif apologies the approach was actually more involved than I had realized - the previous commit was not compiling, my IDE must have been seriously lagging for me to not notice before pushing up.

This does appear to work for our purposes. The problem is that it causes a conflict with the mocked type. I don't know how to get around that conflict yet unfortunately. LMK if you have any ideas.

Reply:
Pushed another commit that manages the conflict